blog/2022/04/simple-chains.md

```diff
@@ -0,0 +1,489 @@
+using SimpleChains, MLDatasets
+
+lenet = SimpleChain(
+  (static(28), static(28), static(1)),
+  SimpleChains.Conv(SimpleChains.relu, (5, 5), 6),
+  SimpleChains.MaxPool(2, 2),
+  SimpleChains.Conv(SimpleChains.relu, (5, 5), 16),
+  SimpleChains.MaxPool(2, 2),
+  Flatten(3),
+  TurboDense(SimpleChains.relu, 120),
+  TurboDense(SimpleChains.relu, 84),
+  TurboDense(identity, 10),
+)
+
+# 3d and 0-indexed
+xtrain3, ytrain0 = MLDatasets.MNIST.traindata(Float32);
+xtest3, ytest0 = MLDatasets.MNIST.testdata(Float32);
+
+xtrain4 = reshape(xtrain3, 28, 28, 1, :);
+xtest4 = reshape(xtest3, 28, 28, 1, :);
+
+ytrain1 = UInt32.(ytrain0 .+ 1);
+ytest1 = UInt32.(ytest0 .+ 1);
+
+lenetloss = SimpleChains.add_loss(lenet, LogitCrossEntropyLoss(ytrain1));
+
+# initialize parameters
+@time p = SimpleChains.init_params(lenet);
+
+# initialize a gradient buffer matrix; number of columns places an upper bound
+# on the number of threads used.
+G = similar(p, length(p), min(Threads.nthreads(), (Sys.CPU_THREADS ÷ ((Sys.ARCH === :x86_64) + 1))));
+
+# train
+@time SimpleChains.train_batched!(G, p, lenetloss, xtrain4, SimpleChains.ADAM(3e-4), 10);
+
+# assess training and test loss
+SimpleChains.error_mean_and_loss(lenetloss, xtrain4, p)
+SimpleChains.error_mean_and_loss(lenetloss, xtest4, ytest1, p)
+
+# train without additional memory allocations
+@time SimpleChains.train_batched!(G, p, lenetloss, xtrain4, SimpleChains.ADAM(3e-4), 10);
+
+# assess training and test loss
+SimpleChains.error_mean_and_loss(lenetloss, xtrain4, p)
+SimpleChains.error_mean_and_loss(lenetloss, xtest4, ytest1, p)
+```
+
+#### PyTorch
+
+Before we show the results, let's look at the competition. Here's two runs of 10 epochs using PyTorch following [this script](https://github.com/chriselrod/LeNetTorch) on a V100 GPU using a batch size of 2048 produces:
+
+```
+Took: 17.33
+Accuracy: 0.9492
+Took: 17.22
+Accuracy: 0.9751
+```
+PyTorch using an AMD EPYC 7513 32-Core Processor:
+```
+Took: 14.86
+Accuracy: 0.9626
+Took: 15.09
+Accuracy: 0.9783
+```
+PyTorch using an Intel i9 10980XE 18-Core Processor:
+```
+Took: 11.24
+Accuracy: 0.9759
+Took: 10.78
+Accuracy: 0.9841
+```
+
+#### Flux.jl
+
+The standard machine learning library in Julia, [Flux.jl](https://github.com/FluxML/Flux.jl) was benchmarked using [this script]() with an A100 GPU. How fast was that?
+
+```julia
+julia> @time train!(model, train_loader)
+ 74.678251 seconds (195.36 M allocations: 12.035 GiB, 4.28% gc time, 77.57% compilation time)
+
+julia> eval_loss_accuracy(train_loader, model, device),
+       eval_loss_accuracy(test_loader, model, device)
+((loss = 0.1579f0, acc = 95.3583), (loss = 0.1495f0, acc = 95.54))
+
+julia> @time train!(model, train_loader)
+  1.676934 seconds (1.04 M allocations: 1.840 GiB, 5.64% gc time, 0.63% compilation time)
+
+julia> eval_loss_accuracy(train_loader, model, device),
+       eval_loss_accuracy(test_loader, model, device)
+((loss = 0.0819f0, acc = 97.4967), (loss = 0.076f0, acc = 97.6))
+```
```

Comment:
Would help for understanding/readability if made clear that the first run includes compile time.  It is kind of mentioned in the summary but readers may get confused if not mentioned earlier.

Reply:
I'll update these examples tonight.
1. Timings will change to reflect the latest changes.
2. I'll add a note about compile time, and also try to improve SimpleChains'.
3. I'll switch the acc, loss order for the Flux benchmarks to reflect the order for SimpleChains and PyTorch.

Reply:
I'll also add a small mlp example.